### PR TITLE
Cleanup ExecutionSystem

### LINKED ISSUE #24180: Executions need cleanup
https://github.com/space-wizards/space-station-14/pull/24150 still needs reviewing e.g.
- It needs prediction.
- It has a bunch of unnecessary null suppression.
- It copies shooting code.

Anything else this is just what I saw at a glance.

### PR DESCRIPTION
**Pull request by nikthechampiongr**
Resolves #24180

## About the PR

This prs cleans up the ExecutionSystem.

Additionally it has the following differences functionally from the current one:
- No doafter modifier based on swing speed on melee weapons.
- Laser weapons can currently not execute due to a limitation in the gunsystem that may be fixed at a later time.(Maybe in this pr if I have time tomorrow).

## Why / Balance

On the no extra damage modifier, I do not think that melee weapons should get a doafter modifier considering in an execution they will do 9 times damage. This can be altered back to its original condition if needed.

## Technical details

- The system was moved to Shared and as much as possible is now predicted.

- Some functions were generalized or removed to reduce duplication.

- All gun code is gone. It now just listen to some events made by the gun system.

- All melee code gone. It now just listens to an event made by the melee system.

- The ExecutionComponent is now added to all weapons that can execute. Using it you can modify values for the doafter, damage modifier, and the many different messages that may be shown on executions.

- Swords got a base type as to not put the component on every single sword individually.

- The ActiveExecutionComponent is added to guns just before they fire and removed right after in order to display the correct messages if the gun fires and to adjust the projectile if the person is shooting themselves so it can collide with them. Similar use for melee weapons.

- The ExecutionProjectileComponent is added to projectiles as they fire. This is later used to make sure that people in the way of an execution don't get 9 times damage as well as to make sure that a clumsy clown shooting their foot doesn't either.

Note on strange thing: In order for me to get shooting yourself to work properly, I create fake coordinates that are slightly offset from the person. This is because otherwise we would be trying to shoot a projectile to its own positions which doesn't work.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

no cl no fun.

### CODE REVIEW

File: Content.Shared/Execution/ActiveExecutionComponent.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/// Added to a gun when it's about to fire an execution shot.
+/// </summary>
+[RegisterComponent, NetworkedComponent]
+public sealed partial class ActiveExecutionComponent : Component
+{
+    [DataField]
+    public EntityUid Attacker;
+
+    [DataField]
+    public EntityUid Victim;
```
Comment (Veritius) [inline Content.Shared/Execution/ActiveExecutionComponent.cs:15]: why are these datafields? this component is only gonna be around for a few seconds

Comment (nikthechampiongr) [reply Content.Shared/Execution/ActiveExecutionComponent.cs:15]: As far as I know I'm supposed to make all fields that will be altered datafield even if it's not really used in stuff like prototypes although I can remove it.

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Rifles/rifles.yml [hunk 2]
```diff
@@ -49,6 +49,15 @@
...
     price: 500
+  - type: Execution
+    suicidePopupInternal: suicide-popup-gun-initial-internal
+    suicidePopupExternal: suicide-popup-gun-external-external
+    suicidePopupCompleteInternal: suicide-popup-gun-complete-internal
+    suicidePopupCompleteExternal: suicide-popup-gun-complete-external
+    executionPopupInternal: execution-popup-gun-initial-internal
+    executionPopupExternal: execution-popup-gun-initial-external
+    executionPopupCompleteInternal: execution-popup-gun-complete-internal
+    executionPopupCompleteExternal: execution-popup-gun-complete-external
```
Comment (Veritius) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Rifles/rifles.yml:60]: it should be generic by default (any gun or sharp melee), with the component overriding the messages.

File: Content.Shared/Execution/ExecutionSystem.cs [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    private void OnCollide(EntityUid uid, ExecutionProjectileComponent comp, StartCollideEvent args)
+    {
+        if (!(args.OtherEntity == comp.Target) || args.OurFixtureId != comp.FixtureId || comp.Clumsy)
+            return;
+
+        if (!TryComp<ProjectileComponent>(uid, out var projectileComponent))
+            return;
+
+        projectileComponent.Damage *= comp.Multiplier;
+    }
```
Comment (metalgearsloth) [inline Content.Shared/Execution/ExecutionSystem.cs:356]: This still has a lot of doubled code in it if it has to manually check clumsy.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
This still seems to double suicide a fair bit so we have 2 systems to maintain doing identical things.

**Comment by nikthechampiongr** [comment]
> This still seems to double suicide a fair bit so we have 2 systems to maintain doing identical things.

I'll check tomorrow if it can even be integrated into the SuicideSystem.

File: Content.Shared/Execution/ExecutionSystem.cs [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+    private bool CanExecuteWithMelee(EntityUid weapon, EntityUid victim, EntityUid user)
+    {
+        if (!CanExecuteWithAny(victim, user))
+            return false;
+
+        // We must be able to actually hurt people with the weapon
+        if (!TryComp<MeleeWeaponComponent>(weapon, out var melee) && melee!.Damage.GetTotal() > 0.0f)
+            return false;
```
Comment (metalgearsloth) [inline Content.Shared/Execution/ExecutionSystem.cs:179]: This is going to cause a null exception, null suppression should be used sparingly.

Comment (nikthechampiongr) [reply Content.Shared/Execution/ExecutionSystem.cs:179]: I thought I had removed all the null suppression code along with that specific if statement which should have its conditions inverted. I guess I accidentally reverted it somewhere? I'll have it removed tomorrow.

File: Content.Shared/Execution/ActiveExecutionComponent.cs (hunk 1 above)
Comment (deltanedas) [reply Content.Shared/Execution/ActiveExecutionComponent.cs]: if the map is saved while executing someone it needs these to persist

Comment (nikthechampiongr) [reply Content.Shared/Execution/ActiveExecutionComponent.cs]: > if the map is saved while executing someone it needs these to persist

I have no idea why the map would save at the exact moment someone is being executed but I'll put them back.

**Comment by metalgearsloth** [comment]
Gonna suss this one out over the weekend sorry on the delay.

**Comment by nikthechampiongr** [comment]
> Gonna suss this one out over the weekend sorry on the delay.

All good thanks. If you know how I could get around the hack of having to create fake entity coordinated to use the gun system for shooting yourself that'd be amazing.

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**